### DISCUSSION TITLE
pick up behavior without slice after setting `parameters` manually

### DISCUSSION THREAD

**Discussion by lukasheinrich** [scikit-hep/awkward#818] [Q&A] [answered]
I'm working on a awkward-array integration with `vector` and `awkward` for `pylhe`

This is useful exercise for building up peharps a more complete event data model for ATLAS as well (@nikoladze @matthewfeickert @kratsg @gordonwatts @alexander-held @henryiii )

The raw arry (ignoring references/links that come layer is this (filtering for briefness)

```python
>>> events = ak.from_iter({
  'particles': [
      {k if k!= 'e' else 'E':getattr(p,k) for k in pylhe.LHEParticle.fieldnames if k in ['px','py','pz','e']} for p in e.particles 
  ],
  'eventinfo':  {k:getattr(e.eventinfo,k) for k in pylhe.LHEEventInfo.fieldnames if k in ['nparticles','weight']}
  }
  
  for e in pylhe.readLHEWithAttributes(skhep_testdata.data_path('pylhe-testfile-pr29.lhe'))
)
>>> ak.type(events)
791 * {"particles": var * {"px": float64, "py": float64, "pz": float64, "E": float64}, "eventinfo": {"nparticles": float64, "weight": float64}}
```

and layout

```python
>>> events.layout.contents[0].content.setparameter('__record__','Particle')
... events.layout.contents[1].setparameter('__record__','EventInfo')
... events.layout.setparameter('__record__','Event')
... ak.type(events)
791 * Event["particles": var * Particle["px": float64, "py": float64], "eventinfo": EventInfo["nparticles": float64]]
```

I can now give names to the records via 
which captures the nesting: 

* The Datatset is an array of events
* The Event consists of EventInfo and an variable array of Particles
* Particles consist of momentum components

registering the following dummy behaviors

```python
class Particle(ak.Record):
    pass

class ParticleArray(ak.Array):
    pass

class Event(ak.Record):
    pass

class EventArray(ak.Array):
    pass

class EventInfo(ak.Record):
    pass

class EventInfoArray(ak.Array):
    pass

ak.behavior['Particle'] = Particle
ak.behavior['*','Particle'] = ParticleArray

ak.behavior['Event'] = Event
ak.behavior['*','Event'] = EventArray

ak.behavior['EventInfo'] = EventInfo
ak.behavior['*','EventInfo'] = EventInfoArray
```

I can get any slice into the array to instantiate

```python
>>> events[:].particles
<ParticleArray [[{px: -0.315, py: -0.63, ... E: 9.35}]] type='791 * var * Partic...'>
```

```python
>>> events[0]
<Event ... nparticles: 9, weight: 0.0751}} type='Event["particles": var * Partic...'>
```

**and** the trivial slice gives me an `EventArray`

```python
>>> events[:]
<EventArray [{particles: [, ... weight: 1.29e-05}}] type='791 * Event["particles...'>
```

** but **  printing out `events` on its own doesn't pick up the toplevel behavior

```python
>>> events
<Array [{particles: [, ... weight: 1.29e-05}}] type='791 * Event["particles": va...'>
```

is there a way to "re-instantiate" the toplevel object so that it also becomes an `EventArray`?

**Comment by lukasheinrich** [comment, marked as answer]
the solution seems to be to either do 

`ak.Array(events)` which should be zero-copy  or to `events[:]` as above

**Reply by jpivarski** [reply]
(I finally read this.) Both `ak.Array(events)` and `events[:]` are zero-copy.

Interestingly enough, `events[:,]` is not zero-copy—the zero-copyness is an implementation detail that is currently different between a pure range slice and a one-element tuple containing a range slice (unless I've already addressed that and just don't remember). So although `events[:]` is not going to be turned from a zero-copy operation to one that does a copy, `ak.Array(events)` is a cleaner way to express the intention.

(Particularly since for Python lists, for which the copy/no-copy distinction is visible beyond performance, as Python lists are mutable, the syntax `some_list[:]` is _supposed to_ copy the list.)

**Reply by lukasheinrich** [reply]
thanks - I'll mark this as an answer!

**Reply by jpivarski** [reply]
Oh, I read this, but I didn't read it carefully enough: don't use `setparameters` directly, as there's an ancient issue to remove it: #177. Use the high-level function [ak.with_parameter](https://awkward-array.readthedocs.io/en/latest/_auto/ak.with_parameter.html), which would give you a new object anyway.

I had misread this, thinking that you changed the `ak.behavior` after creating the array, which is another way you can get into this situation.

**Reply by lukasheinrich** [reply]
how would I achieve the scenario above where I need to set multiple parameters at different nesting levels?

**Reply by jpivarski** [reply]
For something as complex as this, the easiest way would be to build the whole structure from [ak.layout.Content](https://awkward-array.readthedocs.io/en/latest/ak.layout.Content.html) subclass instances, which each have a `parameter` argument to their constructors (instead of building the thing and then having to dig down to find it again).

Also, you should be aware that [ak.from_iter](https://awkward-array.readthedocs.io/en/latest/_auto/ak.from_iter.html) is the slow path, especially if the data start as arrays (I suppose they don't in your case because you have to parse LHE files as text).

I think you want to have 6 growable arrays (a class that allocates a NumPy array and a pointer to "next entry," fills that array until it is full, then allocates and copies data to a new NumPy array that is 1.5× larger):

   * entry offsets: integers starting with 0 (one longer than the number of events), collecting a cumulative sum of the number of particles per event
   * px, py, pz, E, all as flat (no event boundaries) floating-point arrays,
   * weight as a floating-point array (no need for nparticles, as users can get that using [ak.num](https://awkward-array.readthedocs.io/en/latest/_auto/ak.num.html)).

Then you can put these together as flat (no event boundaries) record arrays with the [ak.layout.RecordArray](https://awkward-array.readthedocs.io/en/latest/ak.layout.RecordArray.html) constructor and break that into lists by wrapping it in an [ak.layout.ListOffsetArray64](https://awkward-array.readthedocs.io/en/latest/ak.layout.ListOffsetArray.html). Adding the eventinfo is another [ak.layout.RecordArray](https://awkward-array.readthedocs.io/en/latest/ak.layout.RecordArray.html) around that. By building it up in pieces, you'll have an opportunity to set the parameters at each stage.

Having said all of that, I just noticed that you can also use [ak.ArrayBuilder](https://awkward-array.readthedocs.io/en/latest/_auto/ak.ArrayBuilder.html), passing a `name` argument to the [begin_record](https://awkward-array.readthedocs.io/en/latest/_auto/ak.ArrayBuilder.html#ak-arraybuilder-begin-record) or [record](https://awkward-array.readthedocs.io/en/latest/_auto/ak.ArrayBuilder.html#ak-arraybuilder-record) method. That would certainly be simpler, though [ak.ArrayBuilder](https://awkward-array.readthedocs.io/en/latest/_auto/ak.ArrayBuilder.html) is the performance cost hidden inside [ak.from_iter](https://awkward-array.readthedocs.io/en/latest/_auto/ak.from_iter.html) (it has to iterate over Python objects—that's the problem, which might not matter for you because you're coming from text files, anyway).

**Reply by lukasheinrich** [reply]
yeah the speed is probably limited by text IO/XML parsing  but this seems to work nicely

```python
builder = ak.ArrayBuilder()
for e in pylhe.readLHEWithAttributes(skhep_testdata.data_path('pylhe-testfile-pr29.lhe')):
    with builder.record(name = 'Event'):
        builder.field('eventinfo')
        with builder.record(name =  'EventInfo'):
            for fname in e.eventinfo.fieldnames:
                builder.field(fname).real(getattr(e.eventinfo,fname))
        builder.field('particles')
        with builder.list():
            for p in e.particles:
                with builder.record(name = 'Particle'):
                    builder.field('px').real(1.2)
arr = builder.snapshot()
arr
```
